std-bits/base/src/main/java/org/enso/base/cache/LRUCache.java

```diff
@@ -0,0 +1,311 @@
+    var key = toRemove.getKey();
+    var value = toRemove.getValue();
+    cache.remove(key);
+    lastUsed.remove(key);
+    removeCacheFile(key, value);
+  }
+
+  /** Remove a cache file. */
+  private void removeCacheFile(String key, CacheEntry<M> cacheEntry) {
+    boolean removed = cacheEntry.responseData.delete();
+    if (!removed) {
+      logger.log(Level.WARNING, "Unable to delete cache file for key {0}", key);
+    }
+  }
+
+  /** Remove least-recently used entries until there is enough room for a new file. */
+  private void makeRoomFor(long newFileSize) {
+    removeStaleEntries();
+
+    long totalSize = getTotalCacheSize() + newFileSize;
+    long maxTotalCacheSize = getMaxTotalCacheSize();
+    if (totalSize <= maxTotalCacheSize) {
+      return;
+    }
+
+    // Remove least-recently used entries first.
+    var sortedEntries = getSortedEntries();
+    var toRemove = new ArrayList<Map.Entry<String, CacheEntry<M>>>();
+    for (var mapEntry : sortedEntries) {
+      if (totalSize <= maxTotalCacheSize) {
+        break;
+      }
+      toRemove.add(mapEntry);
+      totalSize -= mapEntry.getValue().size();
+    }
+    assert totalSize <= maxTotalCacheSize;
+    removeCacheEntries(toRemove);
+  }
+
+  private SortedSet<Map.Entry<String, CacheEntry<M>>> getSortedEntries() {
+    var sortedEntries = new TreeSet<Map.Entry<String, CacheEntry<M>>>(cacheEntryLRUComparator);
+    sortedEntries.addAll(cache.entrySet());
+    return sortedEntries;
+  }
+
+  /** Remove least-recently used entries until the total cache size is under the limit. */
+  private void removeFilesToSatisfyLimit() {
+    makeRoomFor(0L);
+  }
+
+  private long getTotalCacheSize() {
+    return cache.values().stream().collect(Collectors.summingLong(e -> e.size()));
+  }
+
+  private long getMaxFileSize() {
+    return maxFileSizeOverrideTestOnly.orElse(maxFileSize);
+  }
+
+  private long getMaxTotalCacheSize() {
+    return maxTotalCacheSizeOverrideTestOnly.orElse(maxTotalCacheSize);
+  }
+
+  public int getNumEntries() {
+    return cache.size();
+  }
+
+  public List<Long> getFileSizesTestOnly() {
+    return new ArrayList<>(
+        cache.values().stream().map(CacheEntry::size).collect(Collectors.toList()));
+  }
+
+  public void setNowOverrideTestOnly(ZonedDateTime nowOverride) {
+    nowOverrideTestOnly = Optional.of(nowOverride);
+  }
+
+  public void clearNowOverrideTestOnly() {
+    nowOverrideTestOnly = Optional.empty();
+  }
+
+  public void setMaxFileSizeOverrideTestOnly(long maxFileSizeOverrideTestOnly_) {
+    assert maxFileSizeOverrideTestOnly_ <= maxFileSize;
+    maxFileSizeOverrideTestOnly = Optional.of(maxFileSizeOverrideTestOnly_);
+  }
+
+  public void clearMaxFileSizeOverrideTestOnly() {
+    maxFileSizeOverrideTestOnly = Optional.empty();
+  }
+
+  public void setMaxTotalCacheSizeOverrideTestOnly(long maxTotalCacheSizeOverrideTestOnly_) {
+    assert maxTotalCacheSizeOverrideTestOnly_ <= maxTotalCacheSize;
```

Comment:
I think this could be a normal check and not just an assert. To disallow users from abusing these test methods (note that assertions will usually be disabled on user-facing builds)

Reply:
Done. I did not convert the exception to an Enso error because this isn't meant to be user-facing.